File: Python/group_anagram.py
'''
Given an array of strings strs, group the anagrams together. You can return the answer in any order.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

Example:
Input: strs = ["eat","tea","tan","ate","nat","bat"]
Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
'''

class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        from collections import defaultdict

        anagrams = defaultdict(list)
        for s in strs:
            anagrams[str(sorted(s))].append(s)

        return list(anagrams.values())

n=["eat", "tea", "tan", "ate", "nat", "bat"]
ans=Solution()
res=ans.groupAnagrams(n)
print(res)

File: C++/Group Anagram.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        map<string, vector<int>> mpp;
        
        for(int i=0;i<strs.size();++i){
            string tmp=strs[i]; 
            sort(tmp.begin(), tmp.end());
            mpp[tmp].push_back(i);
        }
        
        vector<vector<string>> ans(mpp.size());
        int i=0;
        for(auto z: mpp){
            for(auto x: z.second){
                ans[i].push_back(strs[x]);
            }
            i++;
        }
        return ans;
    }
};
